### PR TITLE
add timeout for pipestream connect to prevent infinite timeout when s…

### PR DESCRIPTION
**Pull request by damienfarrar**
…erver is not available

- Added a new option to provide a timeout when connecting to a pipestream as I found it was getting stuck when the server stream wasn't available
- Unit test added

### CODE REVIEW

**Comment by kierunb** [comment]
That's is fantastic commit, can't wait for an updated nuget!

File: GrpcDotNetNamedPipes.Tests/Helpers/NamedPipeChannelContextFactory.cs [hunk 1]
```diff
@@ -22,6 +22,7 @@ namespace GrpcDotNetNamedPipes.Tests.Helpers
     public class NamedPipeChannelContextFactory : ChannelContextFactory
     {
         private readonly string _pipeName = $"GrpcNamedPipeTests/{Guid.NewGuid()}";
+        private readonly int _connectionTimeout = 100;
```
[the remark this answers is not in the document]
Comment (damienfarrar) [reply GrpcDotNetNamedPipes.Tests/Helpers/NamedPipeChannelContextFactory.cs]: Noted, will update

File: GrpcDotNetNamedPipes/Internal/ServerStreamPool.cs [hunk 2]
```diff
@@ -96,6 +96,11 @@ private void ConnectionLoop()
             int fallback = FallbackMin;
             while (true)
             {
+                if (_cts.IsCancellationRequested)
```
[the remark this answers is not in the document]
Comment (damienfarrar) [reply GrpcDotNetNamedPipes/Internal/ServerStreamPool.cs]: I found this was necessary, as the loop would run until the force killing the process in my wpf app. The behaviour can be reproduce by calling `Start` on the NamedPipeServer and then `Kill` it. If you set a breakpoint in the loop you will find it is still running which was causing my wpf app to remain running after killing the NamedPipeServer and then calling Exit.

File: GrpcDotNetNamedPipes/NamedPipeChannel.cs [hunk 3]
```diff
@@ -59,15 +59,23 @@ public NamedPipeChannel(string serverName, string pipeName, NamedPipeChannelOpti
...
                 stream.Dispose();
-                throw;
+
+                if (ex is TimeoutException)
+                {
+                    throw new RpcException(new Status(StatusCode.Unavailable, ex.Message));
+                }
+                else
+                {
+                    throw new RpcException(Status.DefaultCancelled);
```
[the remark this answers is not in the document]
Comment (damienfarrar) [reply GrpcDotNetNamedPipes/NamedPipeChannel.cs]: You are right, I will change it to `throw;`

File: GrpcDotNetNamedPipes.Tests/GrpcNamedPipeTests.cs [hunk 4]
```diff
@@ -366,5 +366,14 @@ private void AssertHasMetadata(Metadata expected, Metadata actual)
...
             }
         }
+
+        [Theory]
+        [ClassData(typeof(MultiChannelClassData))]
+        public void ConnectionTimeout(ChannelContextFactory factory)
+        {
+            var client = factory.CreateClient();
+            var exception = Assert.Throws<RpcException>(() => client.SimpleUnary(new RequestMessage { Value = 10 }));
+            Assert.Equal(StatusCode.Unavailable, exception.StatusCode);
```
[the remark this answers is not in the document]
Comment (damienfarrar) [reply GrpcDotNetNamedPipes.Tests/GrpcNamedPipeTests.cs]: Noted, will update

**Comment by cyanfish** [comment]
Merged, sorry for the delay. I just published version 1.1.0 on Nuget.

**Pull request merged**